documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+## Discussions about the objectives
+### Implementing Fortran and f18 extension semantics 
+Rules to evaluate expressions are listed in Fortran 2018 section 10. For what matters to lowering to FIR, one of the most important part is section 10.1.4 regarding expression evaluation and 10.1.7 regarding evaluation of operands.
+One of the take-away is that Fortran standard gives a lot of freedom to the compiler regarding the order of evaluations of operands and whether they are evaluated or not. The compiler is also allowed to perform mathematically equivalent rewriting of expressions (see 10.1.5.2.4 point 3.).
+A few rules to restrain this freedom during expression lowering to FIR and keep the roles of different compilation phases clear are listed below.
+#### Preserving numerical properties
+##### Rule The lowering of expression to FIR will not perform any expression rewriting while lowering that could lead to different numerical results than the one specified by the semantic context.
+##### Rationale
+Numerical stability is important for many Fortran coders and some others valuate performance more. Options will be required to control numerical stability requirements during compilation. Starting to do a lowering that may change numerical properties will lead to augmenting the number of possible paths in expression lowering to respect these numerical options. Instead, these transformations should be perform directly on `evaluate::Expr` or later on the FIR operations.
+##### Example
+`x + y + z` is not numerically equivalent to `z + x + y`, although it is mathematically.  Lowering to FIR shall not modify the order of the additions in such case.
+##### Plans to enforce the rule while lowering to FIR
+
+This mainly translate in preserving the `evaluate::Expr` tree structure in the FIR operation sequence structure. The operands order should be respected as well as the dependencies. In case the lowering of an expression does not introduce the need to store some sub expressions, the dependency tree of the SSA value holding the result value should match the `evaluate::Expr` tree.
+Concretely, because there is not a one to one match between FIR operations and all Fortran operations and intrinsic procedures. Some rewriting must be done while lowering expressions.
+For instance REAL MAX intrinsics might be lowered to `fir.fcmp` comparison and a MLIR `std.select` operation. In such case, the lowering should ensure that choices and combinations of FIR operations match with the floating-point behavior that is guaranteed for `MAX` given the context (that may include where it is in a program and some compile time options).
+Numerical requirements will need to be accessible in such cases so that lowering can select the right operations and/or propagate the requirements as flags on the operations.
+
+#### Scalar operands evaluation in elemental expressions
+##### Rule
+Scalar operands in elemental expressions must be evaluated only once.
```

Comment:
Can this be extended to `pure` functions?

Reply:
Do you mean, if we have `x = foo() + A + foo()` where we know `foo` is `pure`, we should only call `foo` once for the whole expression and not twice ?
If so, you are correct we can remove redundant calls to pure procedure if we can prove they have the same arguments.
However, to me it is an optimization and we should not attempt to do this while lowering. FIR still has knowledge of procedure `pure` attribute. I think Eric already has an experimental pass to remove these duplicate calls.
So lowering the proposed lowering would still lower `foo() + A + foo()` to:
```
%1 = fir.call foo
%2 = fir.call foo
fir.loop %1 = 1 to n {
  %3 = fir.extract_value %a %i
  %4 = fir.addf %1 %3
  %5 = fir.addf %4 %2
  %6 = fir.coordinate_of %res %i
  fir.store %5 to %6
}
```
And the optimizations would be responsible to remove the redundant call to `foo`.

The idea is to keep the lowering as simple as possible if we know the optimization will later be applied on the generated FIR. Here, a later pass would anyway need to be done to remove redundant pure calls in-between expressions.

Reply:
Re: "an experimental pass"

That is correct. There is already a CSE pass that merges duplicate pure function calls.